### ISSUE TITLE
A critical view on patching os.waitpid() and on guarantees about compatibility with multiprocessing

### ISSUE THREAD

**Issue by jgehrcke** [gevent/gevent#622]
Upfront: this is a complex topic and I just would like to express my (hopefully qualified) doubts. I really do not want to appear as "ranting", so please do not interpret my comment as "destructive", but as an attempt to contribute to gevent's quality assurance and future success.

I want to stimulate a discussion. My main message here is that a lot of thought on specification, desired behavior, and possible side-effects **must** be made before gevent by default monkey-patches `os.waitpid()` in a way that its internal behavior adds considerable complexity on top of the system call itself, or even in a way that its internals do not even invoke the corresponding system call.

Disclaimer: I am the author of [gipc](https://bitbucket.org/jgehrcke/gipc/src/tip/gipc/gipc.py) and have dealt a lot with the gevent/multiprocessing topic, including libev's child watchers and race conditions between Python modules calling `os.waitpid()` vs. libev calling `waitpid()`.

## Short version
gevent 1.1a introduced a patched version of  `os.waitpid()` and I don't think that enough thought went into corresponding specification of behavior. The subprocess module could go without patching `os.waitpid()`, and the multiprocessing module is still "broken" in the context of gevent, even with the patched `os.waitpid()`. That's why I think there is no strong enough reason to monkey-patch `os.waitpid()` in gevent 1.1. I believe this should be reverted, for preventing difficulties in the future. Once introduced, people will rely on it, and I think this could be especially dangerous in this case.

### 1) `os` components should be patched only for a *strong* reason
I think we all agree that Python developers expect the methods provided by the `os` module to be **tight** wrappers around the underlying system calls. These methods provide a way for a developer to invoke a system call easily from within a Python program. These methods are usually only used by developers who know what they are doing, and they usually want to invoke the *underlying system call*. System calls have well-specified behavior. This behavior should be retained by the `os` methods as much as in any way possible, even in a gevent-monkeypatched application.

This leads to a point we should not underestimate: I think we all know that gevent has its reputation issues w.r.t. monkey-patching, and so we really have to make sure that the patched components are patched for a *strong* reason.

### 2) The commit in question 
In [this commit](https://github.com/gevent/gevent/commit/e0d9b60a3179bc7cf0391430d5592cd7b9295f02), Jason made some nice work for making use of libev's child watchers. He also introduced a monkey-patched version of `os.waitpid()`, where a considerable amount of complexity was added on top of the original `os.waitpid()`.

The reason for this drastic override is complex, and it has been motivated by, among others, these issues:
https://github.com/gevent/gevent/issues/600
https://github.com/gevent/gevent/issues/452

Put in very short words, this is mainly about making gevent "somehow" more compatible with Python's multiprocessing (we'll see what that means or doesn't mean further below). 

### 3) Make gevent compatible with multiprocessing -- a tough one

And now we're at the core of my message: we cannot just "easily" make gevent compatible with multiprocessing or parts of multiprocessing. I think we only have two viable options:

1. Developers using gevent need to have specific/precise documentation of how multiprocessing behaves in the context of gevent,
 **or**
2. the gevent documentation must state that multiprocessing is not supported.

Both would be easy to understand for developers, and *most imporantly* the responsibilities would be clarified!

If we opt for (1), **there needs to be thorough and complete specification about how  multiprocessing SHOULD behave in the context of gevent before we even start to release (or even write) corresponding code.**

### 4) Current specification

I do not know what the current maintainers have written down in their notebooks, so I can only refer to publicly available information. Currently there is not a lot of documentation on the gevent/multiprocessing topic in the gevent docs. But there is one statement in the current `whatsnew_1_1.rst` document:

> In addition, simple use of :class:`multiprocessing.Process` is now possible in a monkey patched system, at least on POSIX platforms.

I really appreciate Jason's work here, but clearly, this statement does not fulfill what's written in bold above ("...  thorough and complete specification ...") :-). This is my neutral observation: this statement promises *too much*, and *too vaguely*.

What does it really mean, specifically? That `multiprocessing.Process` can be used within a gevent application? Yes, it can be used, but it still has drastic side-effects:

### 5) Implementation easily violates "specification". An example

It's not difficult to find an example of "simple use of multiprocessing.Process" which is "broken" with gevent 1.1b. MWE:

```python
import gevent
from gevent import monkey
monkey.patch_all()
import multiprocessing


def regularprint():
    while True:
        print("That's just me.")
        gevent.sleep(0.2)


def child():
    gevent.sleep(3)


p = multiprocessing.Process(target=child)
p.start()
g = gevent.spawn(regularprint)
p.join() # This blocks non-cooperatively.
g.kill()
g.join()
```

Execution on gevent 1.1b / Python 3.4.3 / Linux: The parent takes three seconds to execute, but "That's just me." is never printed. The `join()` method blocks non-cooperatively. And remember, this is **with** the monkey-patched `os.waitpid()` in place.

### 7) It's all about (un)certainty

Without any offense towards Jason's work, I'm just asking neutrally: did this patch make gevent and multiprocessing more compatible?

That's difficult to answer, which already is a bad sign. One possible answer is: "to a certain degree, but the details are difficult to explain"

And I think we all agree that a statement like this implicates an amount of uncertainty that engineers do not like to work with, for a reason :-)

### 8) What's required for making gevent "compatible" with multiprocessing?

In my opinion, the test case presented in (5) **has to pass** (in the sense that a context switch to the `regularprint` greenlet is happening upon calling `p.join()`) before we could call multiprocessing.Process even remotely compatible with gevent.

So, can we just make a quick test case from the above's example and implement a quick fix? Sure, but even then I'd say we still cannot call multiprocessing.Process compatible with gevent: as of my work on gipc I am very certain that there are so many corner case issues that *a lot more* test cases will be required!

So... we need more test cases, how about that? The **exact set of required test cases** can only be determined if there is thorough (detailed!) specification of how multiprocessing.Process SHOULD behave in the context of gevent, before we can call them compatible with each other. And from my work on gipc I know that these thoughts soon lead to `multiprocessing.Connection` which is based on C code which would need to be re-implemented, and to other mean little challenges.

Having said this, I am sure that the entire multiprocessing module could be monkey-patched /re-implemented in a way to make it compatible with gevent. It just is quite a big task which requires proper planning.

These thoughts lead to: monkey-patching/re-implementing multiprocessing would significantly increase the gevent maintenance work for following upstream changes. So what really needs to be answered first is which multiprocessing components should be supported *at all*.

When I thought about gipc, I very quickly decided that it should only provide a small sub-set of the functionality of multiprocessing, because otherwise the project would have gotten too big from the very beginning.

# How this affects gevent 1.1

### Should gevent 1.1 make guarantees about compatibility with multiprocessing?

I don't think so.

As described above, I think that's too big for 1.1. 1.1 has other priorities and needs to be finalized soon. I think the official docs for gevent 1.1 should say nothing about multiprocessing or they should state that gevent is incompatible with multiprocessing. We could then define realistic milestones for future gevent versions. That approach would be most future-proof.

### Should gevent 1.1 patch `os.waitpid()`?

I don't think so. 

* For making multiprocessing and gevent "partly" compatible with each other? That should be considered an invalid goal right now, as per the paragraph above.

* For making subprocess & gevent compatible with each other by making use of libev child watchers? I am very sure that this can be done w/o touching `os.waitpid()`:  when using libev child watchers, yes, one needs to make sure that `waitpid()` is *only* called by libev, and therefore never called from within Python code or other imported libraries. Otherwise the application is racy. This is half-easily controllable. In gipc, I have patched multiprocessing's `Popen.poll` for achieving this. In gevent's case this is even easier, because `subprocess` is re-implemented in its entirety anyway.

## Last words

That got lengthy and I really hope I did not offend anyone. I felt urgency to express these thoughts in detail before 1.1 leaves beta status, and really look forward to reading your feedback. 

Cheers,

Jan-Philip

**Comment by parasyte** [comment]
Last month I started a little side-project to monkey patch all parts of the multiprocessing package to make it cooperative. `gipc` was a big source of inspiration, though I couldn't use it directly. I was hoping to clean up my patches and submit a PR here at some point.

In my case, I followed the `gipc` approach and patched `Popen.poll` with a yielding busy-loop that polls the underlying `os.waitpid` with `flags=os.WNOHANG`. @jgehrcke could you elaborate on how "the application is racy" when `os.waitpid` is called from within Python while using libev child watchers? I imagine you mean that two greenlets race one another when one calls `os.waitpid`(e.g. via `Popen.poll`) and the other uses a libev child watcher. But I'm not clear on how this is inherently troublesome.

I can also attest that multiprocessing support with gevent is still quite incomplete. At a bare minimum, `SemLock` and `Connection` need to be patched to allow cooperative locks and queues between processes. This is a deeper discussion for another ticket. (FWIW, the test case in `#5` above works with the additional patching that I do... Along with a more ridiculous test using IPC between a thread and a multiprocessing child.)

In part, I see the draw to patching `os.waitpid`; patching at the lowest level (where the actual blocking operation lies) ensures that all callers will get cooperative behavior. Likewise I agree that the implications of overriding a syscall are unknown at best. Apart from that, I'm indifferent to how or where child polling is patched for support with multiprocessing. Whether it should be patched by default for gevent 1.1 is a topic up for debate, IMHO.

Given that the example in #600 is an incompatibility between gevent and multiprocessing, I don't see that ticket as a strong reason to patch `os.waitpid` by default. #452 mentions `gevent.subprocess`, however it also explicitly calls `os.waitpid`, which seems like The Wrong Way to wait for a "gevent subprocess". I would rather see `gevent.subprocess` implement a cooperative `wait` method, with the caveat that it's The Right Way to wait on a subprocess.

**Comment by jgehrcke** [comment]
> @jgehrcke could you elaborate on how "the application is racy" when os.waitpid is called from within Python while using libev child watchers?

Well, you asked. I'm happy to follow up on this, but note that since you asked for *understanding this*, this is a complex answer which might, but hopefully doesn't, distract others from the actual topic in this thread :-)).

Let's start by looking at a quote from the libev docs:

> "The default loop is the only loop that can handle ev_child watchers, and to do this, it always registers a handler for SIGCHLD."

Mark's words here are to be taken with precision. **"always"**. gevent makes use of this and deviates from this behavior at the same time -- in its `gevent_ev_default_loop()` function defined in https://github.com/gevent/gevent/blob/master/gevent/libev.h, where the crucial line is `sigaction(SIGCHLD, &tmp, &libev_sigchld);` gevent restores the "default" (previous) signal handler and saves libev's SIGCHLD handler safely aside, for potential later usage. The latter becomes restored upon the *first* child watcher registration via gevent API (internally via `gevent_install_sigchld_handler()`, also defined in libev.h). That's important to keep in mind. 

OK, installing a libev child watcher via gevent API re-installs libev's SIGCHLD handler. What does libev do within that handler? First, one needs to recognize the following piece of specification (Single UNIX Specification 3 / POSIX 2001, http://pubs.opengroup.org/onlinepubs/009695399/functions/waitpid.html):

> "A single wait() issued from a SIGCHLD signal handler can be guaranteed to return immediately with status information for a child process."

But what happens when multiple child processes terminate near to "simultaneously"? Linux does not queue signals (in contrast to other platforms): multiple terminating children result in only **one** SIGCHLD signal transmission. So, how can the SIGCHLD handler reliably retrieve information about *all* terminated children up to "now" (until its invocation) or even about those that terminate while it is executing? Simple solution, standard solution, as written down by Michael Kerrisk in The Linux Programming Interface chapter 26.3.1: it, the SIGCHLD handler function, should call waitpid(-1) with the WNOHANG option enabled, in a loop, as often as the return value indicates that a child was reaped (i.e. the loop is to be aborted when the return value indicates that no child terminated). Let's call this approach "SIGCHLD-associated child reap loop". It's an important and established concept. And I will refer to it in the following paragraphs.

A corresponding quote from the libev docs:

> "The occurrence of SIGCHLD is recorded asynchronously, but child reaping is done synchronously as part of the event loop processing. Libev always reaps all children, even ones not watched."

This one is really important to understand, and hard to understand at the same time. "Reaping" means, essentially, reading and deleting kernel state asssociated with the child. It can only be done once, i.e. information about a terminated child can only be retrieved in one place! The docs say that libev always reaps all children. So, why does Python's `os.waitpid()` even work at all when gevent/libev has an event loop running? Look back to the paragraph mentioning `gevent_install_sigchld_handler()`: gevent deactivates this part of libev until the first child watcher is installed.

But what is actually, precisely meant with **"child reaping is done synchronously as part of the event loop processing"**? We'll understand better in a minute, and we'll see how crucial this piece of information is:

libev, as expected, implements the "SIGCHLD-associated child reap loop" concept (https://github.com/gevent/gevent/blob/master/libev/ev.c#L2520) and makes sure that waitpid() is, eventually, called as often as required (at least twice!). However, it doesn't exactly call `waitpid()` within a tight loop. What happens instead, roughly (I'm also no expert here, we should ask Mark): the first call to `waitpid()` happens right away (i.e. the first child zombie is reaped right away), but the next (required!) call to `waitpid()` is scheduled (registered with the event loop) by calling `ev_feed_event()`. And, while the first child is reaped right away, any corresponding child watcher callback is not run directly, but registered with the event loop (also via the mentioned `ev_feed_event()` call). Put simply, the event loop decides when the scheduled functions will be called. I think all this deterministically happens simply within the next event loop interation. The take home message from here is, I guess, that the "SIGCHLD-associated child reap loop" concept becomes tightly integrated with the event loop itself. And that's, I hope and think, exactly what Mark means with "child reaping is done synchronously as part of the event loop processing". That really is an important insight, because: it means that child-watcher associated callbacks are fired *inbetween* two `waitpid()` invocations within the SIGCHLD-associated child read loop and in case of gevent, such a callback usually is a Python function.

**-> essential insight: the CPython interpreter is back in control between two `waitpid()` invocations belonging to libev's SIGCHLD-associated child reap loop.**

Now that's what "child reaping is done synchronously as part of the event loop processing" actually means in the context of Python/gevent. And that has heavy consequences:

When the CPython interpreter is back in control right *within* the SIGCHLD-associated child reap loop, many things can happen. The beginning is clear: the interpreter starts within the child watcher callback, a Python function. Now let's construct a hypothetical scenario. This Python function *might* yield control to the gevent hub which *might* switch to another greenlet which is just about to interact somehow with Python's multiprocessing package. Easily possible, right? Something interesting can happen when this interaction results in, for example, `_cleanup()` being called (this is a function defined in multiprocessing.process: https://github.com/python/cpython/blob/3.4/Lib/multiprocessing/process.py#L52, it is triggered from various different points within the multiprocessing package, for instance during `Process.start()`). This cleanup function in turn calls a certain internal `poll()` method *for every single child registered with the multiprocessing module so far*.

Now guess what happens within that mentioned `poll()` method: Python's `os.waitpid()` is being called, a tight wrapper around `waitpid()`. In this situation there might very well exist kernel state associated with a terminated child process. And libev might sit tight, and desperately want's to retrieve it in the next iteration of its SIGCHLD-associated child reap loop (remember, we're still in there!). But, sad enough, the multiprocessing module right now steals the information. It's lost for good, from libev's point of view.

The termination of the child process whose zombie information was just greedily stolen by multiprocessing (or, generally, by *any other part of the program calling wait/waitpid*) will *never* be recognized by libev. That is, if a child watcher had previously been associated with that process, it will not trigger. And that's exactly what leads to user-facing bugs/unexpected behavior, and so on.

The very hypothetical scenario I have outlined here is exactly what I observed during my tests while working on gipc. I triggered it more easily than I could imagine (by accident, really), ant it was nasty to debug. I have seen in other issues here that Jason also came across this problem and I felt sorry. But once debugged, it's really very simple to understand or at least can be broken down to a very simple concept:

When using the libev child wathcer system it has to be made sure that libev is the only part of a process that calls `waitpid()`. If anything else calls a function from the "wait() family of functions", the libev child watcher system becomes undermined and subject to race conditions(*). It just cant reliably detect child termination anymore.


(*) There are many potential racers involved here, and to understand that this is a race condition we only need to mention two of them: racer A is a terminating process. It is totally unpredictable when it terminates. Racer B is a `wait()/waitpid()` system call not fired off by libev, but implicitly by any greenlet switched to from the gevent hub, while libev is still within a SIGCHLD-associated child reap loop.

And, remember, all this does not apply in the context of gevent until the very first child handler has been installed through gevent API.

In short and simple words: Consider one process that runs libev code and "other code". When this other code invokes a wait()-like system call, libev can easily miss a "this child just terminated" notification. This is non-deterministic behavior (a race condition is involved). When libev misses this notification, a corresponding child watcher callback will never fire, even though the child process has already terminated. So, if one decides to use libev child watchers in a process, one must somehow guarantee that no other code invokes a wait()-like system call.

Closing remarks: in order to understand how to make gevent "work" across forks, and how to properly integrate libev child watchers with gevent, one really needs to dig deeply into the libev documentation, sometimes into the POSIX standard, and it's also important to read up on how *exactly* gevent makes use of libev (by reading code), and sometimes one needs to look into libev code or even consider nasty operating system specifics. I really recommend opening http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod and search for every single occurrence of "child" and "fork". You'll get an impression of what the sensible topics are. On top of that, a really good reference for signal handling and especially for dealing with SIGCHLD is Michael Kerrisk's The Linux Programming Interface.

**Comment by parasyte** [comment]
So the moral of the story is ... gevent should patch `os.waitpid` as part of its child signal handler installation? If I read correctly, that seems to be the gist of it; once libev is asked to do the work, protect libev's assumptions from potentially being destroyed by user code.

Should gevent also patch `signal.signal` and similar calls to further protect itself from the user? I would call this the "sledge hammer" approach to squelching race conditions.

**Comment by jgehrcke** [comment]
>  gevent should patch os.waitpid as part of its child signal handler installation?  If I read correctly, that seems to be the gist of it;

@parasyte: as outlined in the opening post, I do not think that patching os.waitpid is the right approach, especially for gevent 1.1: gevent.subprocess does not require it right now.

And, for future versions of gevent after 1.1: even when monkey-patching os.waitpid() there are many simple ways for calling a wait() system call within a process that uses libev child watchers. That is, replacing os.waitpid is not the ultimate solution for making libev child watchers work reliably. 

The obvious places where os.waitpid() is called from within the stdlib can be monkey-patched (these are not too many), like one can make gevent.subprocess not call os.waitpid() at all. This makes sure that gevent just "works" in many scenarios.

Besides that, the only entirely safe and future-proof approach for making gevent child process monitoring behave properly in the context of more complex scenarios is, in my opinion, documentation: there must be a warning that informs application developers that they have to make sure that no part of their application other than libev calls a wait()-family system call. 

> Should gevent also patch signal.signal and friends to further protect itself from the user? I would call this the "sledge hammer" approach to squelching race conditions.

I think in general the solution must be a combination of code an documentation, whereas documentation must clarify the responsibilities. There are so many ways to make a gevent application break, and gevent for sure should not purely rely on hunting after these scenarios by fixing them with code, but first of all *specify* its behavior via documentation. So that it's easy to point the finger towards application developers.. :-) That's a balancing act, of course.

**Comment by jamadden** [comment]
Patching `os.waitpid` by default was driven by the Principle of Least Surprise: the patched behaviour is far less surprising than the unpatched behaviour. Without patching `waitpid`, *sometimes* it would work, and *sometimes* it wouldn't. 

There are a number of scenarios that could leave app developers scratching their heads. Here's a common one. The application follows all the documentation and monkey-patches immediately. It then proceeds to successfully use either `os.fork/os.waitpid` or `multiprocessing.Process`. Somewhere down the line, it uses `subprocess`. Suddenly, the previously-working use of `fork/waitpid/Process` *stops* working. Surprise! (Why? The first use of `gevent.subprocess` activates child-watchers, which "take over" `waitpid` rendering it racy-at-best. Before that, though, you have normal platform semantics.)

Monkey-patching `subprocess` is enabled by default in gevent 1.1 and before that was enabled by default by important downstream consumers like gunicorn (so, de-facto, has to be effectively considered to be enabled by default in 1.0). However, this scenario can even happen without monkey-patching; if a developer explicitly added calls to `gevent.subprocess` into an app that wasn't monkey-patched than the non-monkey-patched rest of the system would break, defying expectations that gevent would be confined to its own corner. A typical developer would likely respond by saying, "Oh, ok, we need to monkey-patch the system to be compatible; that's the usual advice." But that doesn't work. Surprise!

The bug reports indicate that these scenarios do bite developers in the real world. It seems too much to ask them all to have a complete understanding of POSIX child semantics, libev watchers, etc, when they just want to use a `Process`. To borrow a quote, "simple things should be simple," and that's a large part of what patching `waitpid` by default is intended to enable.

As far as patching functions in `os` is concerned, `fork` was already being patched. `waitpid` goes hand-in-glove, so I see no reason not to patch it. Many seemingly-trivial wrappers around system calls, like `select.select`, are replaced with more involved wrappers in gevent already, this case is no different.

The documentation is kept deliberately under-specified, leaving room to adjust (tighten) the contract in the future. Specific scenarios that are known to work are tested to prevent regressions, but the user is cautioned that when mixing forking and gevent, "here there be dragons" and they should take appropriate care. Immediately below the quoted section of documentation there is this note:

```
In addition, simple use of :class:`multiprocessing.Process` is now
possible in a monkey patched system, at least on POSIX platforms.

.. note:: All of the above entail forking a child process. Forking
		  a child process that uses gevent, greenlets, and libev
		  can have some unexpected consequences if the child
		  doesn't immediately ``exec`` a new binary. Be sure you
		  understand these consequences before using this
		  functionality, especially late in a program's lifecycle.
		  For a more robust solution to certain uses of child
		  process, consider `gipc`_.
```

I'd be happy to add further explicit cautions to the documentation if there is an appropriate place for them.

Now, other than surprising users less, and making more real-world scenarios work, the big danger I can see in patching `waitpid` by default would be regressions to existing code. The test suite is not currently detecting any such regressions, and, so far, nothing has been reported against 1.1b1. If there are regressions before the final release, we'll obviously try to fix them; if they can't be fixed with this design, then their frequency and severity will have to be weighed against the other scenarios that do work to arrive at a consensus about which is most important. (It might also be prudent to add an escape-hatch to not patch `waitpid` should a serious regression be discovered after release, similar to `aggressive` for the `select` module.)

Let me end by saying I appreciate the constructive and informative comments here. This kind of dialog is helpful.

**Comment by jamadden** [comment]
I wrote:
> the big danger I can see in patching waitpid by default would be regressions to existing code.

I did find such a potential regression. It happens in gunicorn, in code that's not necessarily intended to be monkey-patched. If it preloads an application that does monkey patch, it loses the ability to correctly monitor its children. This can be duplicated with this example, which hangs unless the commented-out line (the escape-hatch I also talked about) is uncommented (or unless ``os=False`` is added to the call to ``patch_all``):

```python
import os
import sys
#os.environ['GEVENT_NOWAITPID'] = 'True'

import gevent
import gevent.monkey
gevent.monkey.patch_all()

pid = None
awaiting_child = [True]

def handle_sigchld(*args):
	awaiting_child.pop()

import signal
signal.signal(signal.SIGCHLD, handle_sigchld)

pid = os.fork()
if not pid:
	# child
	gevent.sleep(0.2)
	sys.exit(0)
else:
	while awaiting_child:
		gevent.sleep(0.01)
	print("Done")
```

Of course, this isn't *exactly* new behaviour. If the preloaded app had happened to invoke ``subprocess``, then gunicorn would find itself in exactly the same boat.

A commenter wrote:
> Should gevent also patch `signal.signal` and friends to further protect itself from the user?

I'm looking into that as a fix. In addition, if the call to ``signal.signal`` in the parent is moved to *after* the fork, then the signal gets delivered and the parent exits. I don't see anywhere that (C)Python makes any attempt to preserve and call a previously-installed C-level handler, so this implies that libev and the python signal module would be fighting each other (this may have been mentioned earlier in the thread).

**Comment by parasyte** [comment]
@jamadden Good catch! I'm using PyPy + gunicorn + gevent 1.1b3, but without the preload. But I could have easily hit the same issue. It feels a bit dangerous living on the bleeding edge, sometimes. ;)

**Comment by denik** [comment]
>> Should gevent also patch signal.signal and friends to further protect itself from the user?

I don't think so. There is a use case to use signal.signal rather than gevent.signal -- to interrupt long running code after a timeout. It is also better to use when you want reaction immediatelly rather than on next loop iteration. For instance, this server uses signal.signal to print out event loop's state before exiting: https://github.com/surfly/geventserver/blob/master/geventserver.py#L689 but it still exits immediatelly (there is no need to wait for next loop iteration, and, in fact, doing so might hang the process, because it's possible that signal was sent because of a timeout because the worker's event loop is starving).

IMO we should only monkey-patch something if we know that some important libraries and applications will work better with this patch. We should not monkey-patch things just to hide the fact that there is event loop running underneath - the more aware app developer of this fact, the better. Regarding os.waitpid() I don't have strong opinion except that people using probably know what are they doing and why (and if they don't - nothing will probably help them).

To prevent surprise, maybe gevent/libev's handling of child watchers should be opt-in? Like 

> gevent.subprocess.enable_child_reaping()  # breaks os.waitpid() but enables subprocess.Popen()

without this line gevent.subprocess.Popen raises an exception ("enable_child_reaping() was not called"). If it is called, os.waitpid() and friends break, but that's explicit and documented in enable_child_reaping() doc. (I don't remember how child watchers work anymore, so not sure if this is practical.)

**Comment by jgehrcke** [comment]
Thanks for your feedback Jason and Denis, and sorry: I meant to reply earlier and in more extent, addressing all points raised by you guys. Instead, what follows now is just a tiny contribution to this discussion:

> To prevent surprise, maybe gevent/libev's handling of child watchers should be opt-in?

**Short:** it already is "opt-in" in the sense that it is deactivated by default and becomes activated once `subprocess` is used (this is how I remember the situation, did not look into the code again). This strategy makes sense and should not be changed, in my opinion.

**Long:** My opinion here is that `subprocess` is "officially supported" by gevent, for a long time already, meaning: it should work out of the box. In every meaningful use case of `subprocess`, a child has to be monitored for termination. Either explicitly by the app developer, or *within* the implementation details of `subprocess`. That is, `gevent.subprocess` **must** provide a way to cooperatively wait for a child process to terminate. This in turn means that it **must** use libev child watchers (for proper async notification, i.e. ignoring polling-based solutions). Do I miss something, is my logic broken? If my logic is not broken, gevent supporting `subprocess` means gevent using libev child watchers.

If what I wrote above is true, "opt-in" support of child watchers *within gevent.subprocess does not make sense*. Importing `gevent.subprocess` (or the first child creation) must activate libev child watchers. In that sense, gevent child watcher support is already "opt-in".

In that sense, I want to remind us that gevent's subprocess module can use libev child watchers **without touching os.waitpid()** (which was the situation until the very recent releases, right?). I think this is perfectly fine, as long as this goes along with a big warning sign in the lines of "Any explicit or implicit use of `os.waitpid()` in this situation will lead to unexpected behavior".

A different thought:
Should we decide *not* to monkey-patch the implementation of `os.waitpid()` (and instead clarify in the documentation that its simultaneous use with libev child watchers, be it implicit or explicit, is forbidden/leads to undefined behavior (I am still quite convinced that we should do this)), we could monkey-patch it in a way to emit a runtime warning, pointing to the corresponding gevent documentation. This would ensure that every application developer running into this scenario (be it on purpose or not) is informed.

**Comment by denik** [comment]
Right, patch_all() already patches subprocess by default, so any kind of explicit opt-in would not fit there. Nevermind.

**Comment by jamadden** [comment]
Sorry, I've been out of town due to a death in the family and haven't had a chance to do much more on this. Just some quick thoughts:

- It seems Not OK to me that using one feature of gevent breaks another apparently unrelated feature in bizarre ways. We should strive to be as correct a drop-in replacement as possible (let's face it, not even most developers read all the documentation; I've done what I can to surface possible issues in more places, but as the study of computer interaction tells us, people are conditioned to ignore warnings and just keep going). This seems like a case where we can be more correct and drop-inable.
- For patching `signal.signal` I was planning only to handle the SIGCHLD case (piggybacking off the child watchers), passing everything else through to the original. Since we *know* that SIGCHLD is worthless when gevent.subprocess is or will be used, this should only serve to come closer to the expected behaviour. (I think it's probably very rare that anyone uses SIGCHLD for asynchronous interruption of themselves, right? gunicorn, for example, coopts HUP, INT, TERM, QUIT, TTIN, TTOU, WINCH, USR1, and USR2, but CHLD is too useful---of course that could be a flawed example since one of the main things gunicorn does is watch children :). If someone is currently using SIGCHLD for non-standard purposes, they obviously can't be using gevent.subprocess, so in the upgrade they'd just have to flip a switch to keep SIGCHLD available, and keep not using/patching gevent.subprocess, which seems fairly reasonable to me---they'd already have to flip a switch to disable patching gevent.subprocess in if they were using subprocess at all). But like I said, I haven't gotten around to writing any code and test cases, which is where flaws in plans like these are usually discovered. Maybe this weekend...
